File: solutions/2026-05-06/python/2026-05-06-rotating-the-box.py
```python
"""
Problem Summary:
Given a 2D grid representing a box with stones, obstacles, and empty spaces,
rotate the box 90 degrees clockwise and simulate stones falling due to gravity.

Link: https://leetcode.com/problems/rotating-the-box/

Approach:
The problem can be solved in three main steps:
1. Rotate the grid 90 degrees clockwise. This means an element at `boxGrid[r][c]`
   will move to `rotatedGrid[c][m-1-r]`.
2. For each column in the rotated grid, simulate gravity. This involves iterating
   from bottom to top. We maintain a pointer to the next available position for a stone.
   When we encounter a stone ('#'), we place it at the current available position
   and increment the pointer. When we encounter an obstacle ('*'), we update the
   available position to be just above the obstacle. Empty spaces ('.') are
   simply skipped as stones will fall into them.
3. Return the grid after gravity simulation.

Time Complexity:
Let m be the number of rows and n be the number of columns in the input grid.
1. Rotation: O(m * n) to create the initial rotated grid.
2. Gravity Simulation: For each column of the rotated grid (which has length m),
   we iterate through its elements. This takes O(m) time per column. Since there
   are n columns, this step is O(m * n).
Overall time complexity is O(m * n).

Space Complexity:
1. Rotated Grid: We create a new grid of size n x m, which takes O(n * m) space.
2. Gravity Simulation: We modify the rotated grid in-place.
Overall space complexity is O(n * m).
"""

class Solution:
    def rotateTheBox(self, boxGrid: list[list[str]]) -> list[list[str]]:
        m = len(boxGrid)
        n = len(boxGrid[0])

        # Step 1: Rotate the grid 90 degrees clockwise
        # The dimensions of the rotated grid will be n x m.
        # An element at boxGrid[r][c] moves to rotated_grid[c][m-1-r].
        rotated_grid = [['.' for _ in range(m)] for _ in range(n)]
        for r in range(m):
            for c in range(n):
                rotated_grid[c][m - 1 - r] = boxGrid[r][c]

        # Step 2: Simulate gravity for each column in the rotated grid
        # Iterate through each column of the rotated grid.
        for c in range(m): # The columns of rotated_grid correspond to rows of original boxGrid
            # 'fall_pos' tracks the next available row index from the bottom where a stone can land.
            # Initialize 'fall_pos' to the last row index of the current column.
            fall_pos = n - 1
            
            # Iterate from the bottom row upwards in the current column.
            for r in range(n - 1, -1, -1):
                cell = rotated_grid[r][c]
                
                if cell == '#':
                    # If we find a stone, place it at the current 'fall_pos'
                    # and move 'fall_pos' one step up for the next stone.
                    if r != fall_pos: # Only move if it's not already in place
                        rotated_grid[fall_pos][c] = '#'
                        rotated_grid[r][c] = '.' # Clear the original stone position
                    fall_pos -= 1
                elif cell == '*':
                    # If we find an obstacle, it stops stones from falling further down.
                    # The next available landing position ('fall_pos') must be just above this obstacle.
                    fall_pos = r - 1
                # If the cell is '.', it's an empty space, stones will fall into it.
                # We don't need to do anything specific for '.' here, as 'fall_pos'
                # already accounts for stones falling through empty spaces.

        # Step 3: Return the grid after gravity simulation
        return rotated_grid

```

File: solutions/2026-05-06/java/2026-05-06-rotating-the-box.java
/**
 * Problem: Rotating the Box
 * Summary: Rotates a 2D grid representing a box containing stones, obstacles, and empty spaces 90 degrees clockwise. Stones fall due to gravity after rotation.
 * Link: https://leetcode.com/problems/rotating-the-box/
 *
 * Approach:
 * 1. Simulate the rotation: A 90-degree clockwise rotation of an m x n grid results in an n x m grid.
 *    The element at boxGrid[i][j] in the original grid will move to rotatedGrid[j][m-1-i] in the new grid.
 * 2. Handle gravity after rotation: For each column in the rotated grid, simulate gravity.
 *    Iterate from bottom to top. Maintain a pointer to the next available empty slot for a stone.
 *    If an obstacle '*' is encountered, it acts as a new bottom for stones above it. Reset the empty slot pointer.
 *    If a stone '#' is encountered, place it in the next available empty slot and update the pointer.
 *
 * Time Complexity:
 * - Rotation: O(m * n), where m is the number of rows and n is the number of columns in the original grid.
 * - Gravity simulation for each column: For each of the n columns in the rotated grid (which has m rows), we iterate through the m rows. So, O(n * m).
 * - Total Time Complexity: O(m * n) + O(n * m) = O(m * n).
 *
 * Space Complexity:
 * - New grid for rotation: O(n * m) to store the rotated grid.
 * - Total Space Complexity: O(n * m).
 */
class Solution {
    /**
     * Rotates the box grid 90 degrees clockwise and simulates gravity.
     *
     * @param boxGrid The input 2D character array representing the box.
     * @return A new 2D character array representing the box after rotation and gravity.
     */
    public char[][] rotateTheBox(char[][] boxGrid) {
        int m = boxGrid.length;
        int n = boxGrid[0].length;

        // 1. Rotate the box 90 degrees clockwise.
        // The new grid will have dimensions n x m.
        // An element at originalGrid[i][j] moves to rotatedGrid[j][m-1-i].
        char[][] rotatedGrid = new char[n][m];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                rotatedGrid[j][m - 1 - i] = boxGrid[i][j];
            }
        }

        // 2. Simulate gravity for each column in the rotated grid.
        // Iterate through each column of the rotated grid.
        for (int j = 0; j < m; j++) { // j iterates through columns of rotatedGrid (which corresponds to original rows)
            // `nextEmptyRow` points to the next available row for a stone in the current column.
            // It starts from the bottom-most row and moves upwards.
            int nextEmptyRow = n - 1; // n is the number of rows in the rotated grid.

            // Iterate from bottom to top for the current column.
            for (int i = n - 1; i >= 0; i--) { // i iterates through rows of rotatedGrid
                char cell = rotatedGrid[i][j];

                if (cell == '*') {
                    // If an obstacle is found, it becomes the new "bottom" for stones above it.
                    // Reset `nextEmptyRow` to the row just above the obstacle.
                    nextEmptyRow = i - 1;
                } else if (cell == '#') {
                    // If a stone is found, move it to the `nextEmptyRow` if it's not already there.
                    if (i != nextEmptyRow) {
                        rotatedGrid[nextEmptyRow][j] = '#';
                        // The original position of the stone becomes empty.
                        rotatedGrid[i][j] = '.';
                    }
                    // Move `nextEmptyRow` one step up for the next stone.
                    nextEmptyRow--;
                }
                // If the cell is '.', do nothing, it remains empty or gets filled by a falling stone.
            }
        }

        return rotatedGrid;
    }
}
